handler/src/main/java/io/netty/handler/ssl/ConscryptAlpnSslEngine.java

```diff
@@ -0,0 +1,173 @@
+        return data == null ? null : new String(data, CharsetUtil.US_ASCII);
+    }
+
+    private static byte[] toProtocolArray(List<String> protocols) {
+        int length = 0;
+        for (int i = 0; i < protocols.size(); ++i) {
+            length += 1 + protocols.get(i).length();
+        }
+        byte[] data = new byte[length];
+        for (int dataIndex = 0, i = 0; i < protocols.size(); ++i) {
+            String protocol = protocols.get(i);
+            int protocolLength = protocol.length();
+
+            // Add the length prefix.
+            data[dataIndex++] = (byte) protocolLength;
+            for (int ci = 0; ci < protocolLength; ++ci) {
+                // Assume US-ASCII.
+                data[dataIndex++] = (byte) protocol.charAt(ci);
+            }
+        }
+        return data;
+    }
+
+    private static final class ClientEngine extends ConscryptAlpnSslEngine {
+        private final ProtocolSelectionListener protocolListener;
+
+        ClientEngine(OpenSSLEngineImpl engine,
+                JdkApplicationProtocolNegotiator applicationNegotiator) {
+            super(engine, applicationNegotiator.protocols());
+            protocolListener = checkNotNull(applicationNegotiator
+                            .protocolListenerFactory().newListener(this, applicationNegotiator.protocols()),
+                    "protocolListener");
+        }
+
+        @Override
+        public void onHandshakeFinished() throws SSLException {
+            String protocol = getAlpnSelectedProtocol(getWrappedConscryptEngine());
+            try {
+                protocolListener.selected(protocol);
+            } catch (SSLException e) {
+                throw e;
+            } catch (Throwable cause) {
+                PlatformDependent.throwException(cause);
+            }
+        }
+    }
+
+    private static final class ServerEngine extends ConscryptAlpnSslEngine {
+        private final ProtocolSelector protocolSelector;
+
+        ServerEngine(OpenSSLEngineImpl engine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+            super(engine, applicationNegotiator.protocols());
+            protocolSelector = checkNotNull(applicationNegotiator.protocolSelectorFactory()
+                            .newSelector(this,
+                                    new LinkedHashSet<String>(applicationNegotiator.protocols())),
+                    "protocolSelector");
+        }
+
+        @Override
+        public void onHandshakeFinished() throws SSLException {
+            try {
+                String protocol = getAlpnSelectedProtocol(getWrappedConscryptEngine());
+                protocolSelector.select(protocol != null ? Collections.singletonList(protocol)
+                        : Collections.<String>emptyList());
+            } catch (SSLException e) {
+                throw e;
+            } catch (Throwable cause) {
+                PlatformDependent.throwException(cause);
```

Comment:
actually you may want to consider being consistent with `JettyAlpnSslEngine` and wrap in a `SSLHandshakeException`.